compiler/src/iree/compiler/Dialect/Flow/Transforms/SplitReduction.cpp

```diff
@@ -76,14 +82,87 @@ struct LinalgSplitReduction
   LinalgExt::LinalgTransformationFilter filter;
 };
 
+/// Sets the split reduction value for topk operation using a simple heuristic.
+/// The goal here is to increase parallelism.
+static int64_t topkSplitReduceRatio(int64_t splitReductionDepth,
+                                    LinalgExt::TopkOp topkOp, int64_t kValue) {
+  // Step 0. If user set the split reduction ratio use it.
+  if (!topkSplitReductionRatio.empty()) {
+    SmallVector<int64_t, 4> reductionRatios(topkSplitReductionRatio.begin(),
+                                            topkSplitReductionRatio.end());
+    if (splitReductionDepth >= reductionRatios.size()) {
+      return kNOTopkSplitReductionRatioDefault;
+    } else {
+      return reductionRatios[splitReductionDepth];
+    }
+  }
+
+  // Step 1. Hard to predict of advantage of splitting reduction more than 2
+  // depth.
+  if (splitReductionDepth > 1) return kNOTopkSplitReductionRatioDefault;
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "\n--- topkSplitReduceRatio started :";
+    llvm::dbgs() << "splitReductionDepth=" << splitReductionDepth
+                 << "\t kValue= " << kValue << "\n";
+  });
+
+  // Step 2. Set split value for for dynamic shape.
+  auto inputType = topkOp.getInputs().front().getType();
+  auto rankedType = dyn_cast<RankedTensorType>(inputType);
+  if (!rankedType) {
+    return kTopkSplitReductionRatioDefault;
+    LLVM_DEBUG({
+      llvm::dbgs() << "--- tensor is not ranked tensor, setting splitk value "
+                      "as default: "
+                   << kTopkSplitReductionRatioDefault << "\n";
+    });
+  }
+
+  // Step 3. Find a value that balances the workload between the two kernels.
+  int64_t lastDim = rankedType.getDimSize(rankedType.getRank() - 1);
+  auto findSplitKValue = [](int64_t workload, int64_t kValue) {
+    for (int64_t ki = 3; ki < kValue; ki++) {
+      if (workload % ki != 0) continue;
+      int workload_2nd_kernel = kValue * ki;
+      int workload_1st_kernel = workload / ki;
+      int check = workload / (ki - 1);
+      LLVM_DEBUG({
+        llvm::dbgs() << "1stKernel = " << workload_1st_kernel << "\t"
+                     << "2ndKernel = " << workload_2nd_kernel << "\t"
+                     << "Check Kernel = " << check << "\n";
+      });
+      if (workload_2nd_kernel > workload_1st_kernel &&
+          workload_2nd_kernel > check) {
```

Comment:
Could you add a comment explaining the logic why this is a good split?
Also do you believe we can keep the logic target independent like this?
For instance, do you think we should have one heuristic for the GPU and one for the CPU?

Reply:
The heuristic attempts to split work as much as evenly between the dispatch regions. It splits with maximum depth of 2. I think it is okay for CPU and GPU.

CPU has single dimensional parallelism (excluding numa), so splitting with 2 depth would be fine. 

GPU has two dimensional parallelism and each dimension has 3 more dimensions (blockIdx x,y,z), so splitting with 2 depth is not the best choice. 

I don't know the target device at this phase. Therefore, I kept the split depth at 2.

Reply:
Okay, that's what I expected: not all targets would like the same splits.
Couple of follow-up questions:
- @grypp if we were to tune that heuristic solely for NV GPU, how much more perf could we get?
- @MaheshRavishankar what is the recommended way to plug this kind of target specific tweaks?

Reply:
Just coming online... The layering of these has been off from the get go. So this should be moved into plugin/pre-processing for now to make quick progress. The more structural issue will require more time (but happy to help).

Reply:
If tune this specifically in nvgpu, I could add additional parallelism, but the benefit would be negligible. This was the case from the cases I tried.

Reply:
@jpienaar how do you think we should approach these target specific heuristics?
For now, we can go crazy in the plugins, but flagging this one in case this sparks some ideas on how we should do things in the platform.

Reply:
Moving to plugin or pre-processing pass for now is POR.

More target specific heuristics should be queryable and here it should just be an invocation of a returned function pointer. The main consideration is 1) what attributes are queryable, 2) interface of predictor, 3) where. Fixing that and then we will do something a la const-eval. I'll write up a one pager here later, but the above in short.